lib/types.js

```diff
@@ -1246,54 +1315,21 @@ UnionType.prototype._branchConstructor = function () {
  * + `map`, `record`
  */
 class UnwrappedUnionType extends UnionType {
-  constructor (schema, opts) {
+  constructor (schema, opts, /* @private parameter */ _projectionFn) {
     super(schema, opts);
 
-    this._dynamicBranches = null;
-    this._bucketIndices = {};
-    this.types.forEach(function (type, index) {
-      if (Type.isType(type, 'abstract', 'logical')) {
-        if (!this._dynamicBranches) {
-          this._dynamicBranches = [];
-        }
-        this._dynamicBranches.push({index, type});
-      } else {
-        let bucket = getTypeBucket(type);
-        if (this._bucketIndices[bucket] !== undefined) {
-          throw new Error(`ambiguous unwrapped union: ${j(this)}`);
-        }
-        this._bucketIndices[bucket] = index;
+    if (!_projectionFn && opts && typeof opts.wrapUnions === 'function') {
+      try {
+        _projectionFn = opts.wrapUnions(this.types);
+      } catch(e) {
+        throw new Error(`Error generating projection function: ${e}`);
       }
-    }, this);
-
-    Object.freeze(this);
-  }
-
-  _getIndex (val) {
-    let index = this._bucketIndices[getValueBucket(val)];
-    if (this._dynamicBranches) {
-      // Slower path, we must run the value through all branches.
-      index = this._getBranchIndex(val, index);
     }
-    return index;
-  }
+    this._getIndex = _projectionFn
+      ? generateProjectionIndexer(_projectionFn)
+      : generateDefaultIndexer.bind(this)(this.types);
```

Comment:
`generateDefaultIndexer(this.types)`

Reply:
d6692e16e5593d8bd48778e336e07a8fd7b8820a